cf-oscore/src/main/java/org/eclipse/californium/oscore/HashMapCtxDB.java

```diff
@@ -59,20 +64,62 @@
 	public HashMapCtxDB() {
 
 		this.tokenMap = new HashMap<>();
-		this.ridMap = new HashMap<>();
+		this.contextMap = new HashMap<>();
 		this.uriMap = new HashMap<>();
 		this.seqMap = new HashMap<>();
 		this.allTokens = new ArrayList<Token>();
 	}
 
+	/**
+	 * Retrieve context using RID and ID Context. If the provided ID Context is
+	 * null a result will be returned if there is only one unique context for
+	 * that RID.
+	 */
 	@Override
-	public synchronized OSCoreCtx getContext(byte[] rid) {
-		if (rid != null) {
-			return ridMap.get(new ByteId(rid));
-		} else {
+	public synchronized OSCoreCtx getContext(byte[] rid, byte[] IDContext) throws CoapOSException {
+		// Do not allow a null RID
+		if (rid == null) {
 			LOGGER.error(ErrorDescriptions.BYTE_ARRAY_NULL);
 			throw new NullPointerException(ErrorDescriptions.BYTE_ARRAY_NULL);
 		}
+
+		HashMap<ByteId, OSCoreCtx> matchingRidMap = contextMap.get(new ByteId(rid));
```

Comment:
Just as exercise for good coding practice: count the `new ByteId(byte[])` and check, if this could be reduced by at least reuse that ByteId within the method. It may get even less used, if you store the rid as ByteId and not as byte array.

Reply:
True, it was used excessively. Thanks for the feedback. I have now made a commit trying to reduce how much it is used, see commit 19f0fc1ebdae42eaff3b70dbedd60866730c2ab7

As for storing the rid differently, perhaps that could be a thing for future improvements?